### ISSUE TITLE
A space character flagging an email as invalid

### ISSUE THREAD

**Issue by Ben-MyBB** [mybb/mybb#3088]
I have just started a Mybb site and with the 50 or so users who have attempted to register, I have run into this issue twice that I know of.  I do classify it as a bug although it may be a design bug and not a coding bug.  Either way it should be addressed.  It is confusing for users and took hours for me to figure out what was going on.

If one enters an email address of the form "xyz@abc.com ", what one sees on the screen is just 
xyz@abc.com".  if in response to typing that, one gets a message that the email address was invalid, the user sees what appears to be a perfectly valid email address as being rejected.  Either the code needs to be corrected to ignore leading/trailing white-space (which is the norm for this type of thing) or the less user-friendly but minimally acceptable solution of having a more descriptive error message that at least echoes the email address that was rejected (eg, "The address you entered, 'xyz@abc.com ', is not a valid email address." or even better "The address you entered, 'xyz@abc.com ', is not a valid email address. Email addresses may not contain the character ' '.")

Original thread: [A space character flagging an email as invalid](https://community.mybb.com/thread-216476.html)

**Comment by labrocca** [comment]
There are many places in MyBB which they don't use trim() in username or email input and I wish they would.  I often copy/paste for a page and if there is any space in the copy and I don't catch it I get an error and have to do the search again.  MyBB should trim the entire input action class imho.  

I think this can all be solved with a trim() in function clean_input() inside inc/class_core.php.

**Comment by effone** [comment]
Totally agree with you. Should not always rely on front end validations, in fact front end and back end validation system should be inline with each other.

But what the OP has said is a kinda not possible unless he has done some mess with the registration page and destroyed front end validation (javascript) as it doesn't allow space character in emails ("me@mysite.com " << Kinda thing) as described and blocks at the first place itself.

Though the front end validation requires improvement as well, logically as if you type the same invalid email in 'Confirm Email' field: it says: _You need to enter the same email address again_ : though they are invalid but same, hence the error message is misguiding.

**Comment by effone** [comment]
So, where is the perfect place to trim?
1. .\inc\datahandler\user.php : where `trim_blank_chrs()` only affects to check blank email which should at first place to be trimmed to affect other checks like invalid and blocked.
2. In the member.php while accumulating user provided data in an array $user, we can impose `trim_blank_chars()` on `$mybb->get_input('email')` as well as `$mybb->get_input('email2')`.
3. Inside `get_input()` method under class_core.php to affect the trim for all input data...

**Comment by labrocca** [comment]
I think imho it should trim all input data. In what case would blank space outside the string be used?  MyBB already trims usernames when they are created and for username changes.  You can't have an email with blank characters on the outside.  I can't think of any situation where a trim for all input would effect anything negatively.  Can anyone else?

**Comment by euantorano** [comment]
Only possibility I can think of is for post/PM data, but realistically when would you want spaces at the start/end of a post?

**Comment by dvz** [comment]
MyBB shouldn't arbitrarily modify input fields -  blank characters may be used in passwords and 3rd party extensions; if there's no reason to trim leading/trailing spaces for specific fields, no need to limit the possibilities for users/developers who might find it useful.

**Comment by labrocca** [comment]
MyBB already trims() the username but I see they don't do it for password. I took the time and tried to create a password with blank spaces at the beginning and end.  It doesn't work.  Example password:

"   7&B2Yb133BFF"  I created an account with that as password. It allowed me to login with it but also without the space as "7&B2Yb13333FF".  So the password field is removing the space at least and probably tab and maybe empty characters.

I just think the best option is to do a trim on all input. Let's say you have someone that did use a blank character in their password, they can still do a password reset and the admin can announce that anyone with blank characters in the beginning or end of their password should change it immediately.

Right now things are half-trimmed.  We trim the input and it's 100% trimmed.  I say go with that.
